Create Namespace registry

oruebel opened the pull request:
This PR is to simplify and clarify the process for how to implement an NWB extension in NWB. In particular this PR revamps how we manage schema/namespaces to simplify the process of integrating new schema. 

## Issues resolved by this PR

Fix #169
Fix #179 
Fix #180

## Tasks
- [X] Created a new `NamespaceRegistry`. The registry is populated statically by the compiler and collects all namespaces used. This allows us to simplify the integration of additional namespaces from extensions and makes the management of namespaces more transparent overall.
- [X] Updated the existing spec header files to use call `REGISTER_NAMESPACE` to register the main namespaces
- [X] Updated the existing spec files to use a ``const std::vector`` instead of a ``constexpr std::array`` to simplify use of different namespaces with different numbers of spec files
- [X] Updated `NWBFile::cacheSpecifications` and `NWBFile::Initialize` to use the new `NamespaceRegistry` directly 
- [X] Refactored the `generate_spec_files.py` script to make it easier to use for others as a command line tool, improve documentation, and simplify configuration for use with extensions 
- [X] Updated  `generate_spec_files.py` to support the new `NamespaceRegistry`
- [X] Add developer docs on how to use the `NamespaceRegistry`
- [X] Updated REGISTERED_TYPE macro to support setting the namespace name via the `const std::string namespaceName` variable from the spec header files

## Optional task (to be addressed in separate PR)
- [X] Create a demo for writing an extension. This has been implemented in PR #183 (issue #182)

oruebel commented:
@stephprince this PR is ready for review

stephprince commented on resources/generate_spec_files.py:
```diff
@@ -1,101 +1,182 @@
+        except Exception as e:
+            logger.error(f"Failed to load YAML file {filepath}: {e}")
+            raise
+    elif filepath.suffix == '.json':
+        try:
+            with open(filepath) as f:
+                return json.load(f)
+        except Exception as e:
+            logger.error(f"Failed to load JSON file {filepath}: {e}")
+            raise
```
```suggestion
```

Since this script now generates the header files directly from the schema .yaml files, I think we no longer need to include the json files or support loading them.

stephprince commented on resources/generate_spec_files.py:
```diff
@@ -1,101 +1,182 @@
+                schema_file = namespace_file.parent / s['source']
+                # If the schem file is missing, check if it was converted to JSON/YAML
+                if not os.path.exists(schema_file):
+                    for ext in ['.yaml', '.yml', '.json']:
```
```suggestion
                    for ext in ['.yaml', '.yml']:
```

See previous comment about supporting the .json files

stephprince commented on resources/generate_spec_files.py:
```diff
@@ -1,101 +1,182 @@
+    """
+    logger.info(f"Starting to process schema files in directory: {schema_dir}")
+    for file in schema_dir.rglob(r"*namespace.*"):
+        if file.suffix in ['.yaml', '.yml', '.json']:
```
```suggestion
        if file.suffix in ['.yaml', '.yml']:
```

See comment above about supporting the json files. I was running into errors when pointing to the .json directory anyway.

stephprince reviewed (commented):
Looks good! 

Most of my comments are related to `generate_spec_files.py` - I think we should update the files in the `src/spec/` folder to be generated by the most recent version of this script. When I regenerated the core and hdmf-common header files, I currently get some validation errors. I think my comments should address most of those but should double check after these updates.

These updates should also allow us to remove the `resources/spec` folder entirely and work directly with the yaml files from nwb-schema.

oruebel commented on resources/generate_spec_files.py:
```diff
@@ -1,101 +1,182 @@
+
+def load_file(filepath: Path) -> Union[Dict, List]:
+    """
+    Load a YAML or JSON file and return its content.
```
```suggestion
    Load a YAML schema file and return its content.
```

oruebel commented:
@stephprince thanks for the helpful suggestion. I have now addressed the remaining issues with `generate_spec_files.py` that you had identified. Please take a look.

oruebel commented:
> These updates should also allow us to remove the `resources/spec` folder entirely and work directly with the yaml files from nwb-schema.

Good idea. I've implemented this change in https://github.com/NeurodataWithoutBorders/aqnwb/pull/199 I created a separate PR because I needed to branch of another PR in order to fix the workflows and pick up all the other required changes to python scripts.
